STAR-410

Gerrrr opened the pull request:
This PR is based on https://github.com/datastax/cassandra/pull/94 and https://github.com/datastax/cassandra/pull/93; the first 2 commits can be ignored.

Changes compared to the internal PR:
* In https://github.com/datastax/cassandra/pull/101/commits/8427634dd8936cb696474f1919612a776c26b1c5 I skipped refactoring done to LCS manifest as it clashes with the changes in OSS and are not required to port UCS.
*  baaa178 reverts pieces of refactoring that were only useful to enable insights.
* In f321673a38b3f171ed87c4383992feec069d9788 `CompactorIterator` keeps `TableOperation` as a part of its state to make possible quick interruption via `AbortableUnfilteredPartitionTransformation`.

blambov commented on src/java/org/apache/cassandra/db/compaction/CompactionStrategyManager.java:
```diff
@@ -1039,6 +1022,14 @@ public String getName()
+    private List<List<AbstractCompactionStrategy>> getStrategies(boolean checkBoundaries)
+    {
+        if (checkBoundaries)
+            maybeReloadDiskBoundaries();
```
This call repeats below. Is this intentional?

Gerrrr replied:
Fixed in db135b8.

blambov reviewed (approved):
I can't see serious problems. However, I think we need to include DB-4711, because it may still cause the problem we saw in DB-4290.

This is a very large change and I really don't know how it will be seen by the OSS community. We should think about clarifying the benefits it brings and possibly bundling it together with the `CompactionStrategyManager` avoidance refactor.

Gerrrr commented on src/java/org/apache/cassandra/db/compaction/AbstractTableOperation.java:
```diff
@@ -215,86 +147,89 @@ public Progress(TableMetadata metadata, OperationType operationType, long comple
-            else
-            {
-                buff.append('(');
+                buff.append(String.format("@%s(%s, %s)", metadata.id, metadata.keyspace, metadata.name));
```
This makes the string the same as it was in CompactionInfo:

https://github.com/datastax/cassandra/blob/5150d26ba2b06c4ae297b2102c9f882e4589ebaf/src/java/org/apache/cassandra/db/compaction/CompactionInfo.java#L131-L144

Gerrrr commented on src/java/org/apache/cassandra/db/compaction/CompactionInfo.java:
```diff
@@ -225,7 +225,7 @@ public StopTrigger trigger()
     }
 
-    public enum Unit
+    public static enum Unit
```
Fixed in https://github.com/datastax/cassandra/pull/101/commits/37afb73102c96cca3ff6758a5feeb10171c085de#diff-64fbdaae22e8fbcd44a85e02176f0ca28530c47db6acaa32bdf4eb410030dd1cR132.

Gerrrr commented on src/java/org/apache/cassandra/db/compaction/CompactionManager.java:
```diff
@@ -1103,38 +1098,30 @@ public void disableAutoCompaction()
         {
-            if (scrubInfo != null)
-                activeCompactions.finish(scrubInfo);
+            Throwables.maybeFail(ex);
```
Fixed in c4b6066 .

Gerrrr commented on src/java/org/apache/cassandra/db/compaction/BackgroundCompactions.java:
```diff
@@ -0,0 +1,259 @@
+    }
+
+    /**
+     * Creates new aggregates with the pending aggregates but adding any existing aggregates with
```
Fixed in 235c85f .

Gerrrr commented on src/java/org/apache/cassandra/db/compaction/BackgroundCompactions.java:
```diff
@@ -0,0 +1,259 @@
+    /** The compaction logger */
+    private final CompactionLogger compactionLogger;
+
+    /** The compaction aggregates wither either pending or ongoing compactions, or both. */
```
Fixed in 235c85f.

Gerrrr commented on src/java/org/apache/cassandra/db/compaction/CompactionTask.java:
```diff
@@ -60,18 +67,93 @@
+    }
+
+    /**
+     * Create a compaction task for tombstone collection.
```
Fixed in 235c85f.

Gerrrr commented:
@blambov I included DB-4711 and addressed your comments.

This is a very large change and I really don't know how it will be seen by the OSS community. We should think about clarifying the benefits it brings and possibly bundling it together with the CompactionStrategyManager avoidance refactor. Can you please take a another look before I merge?

> This is a very large change and I really don't know how it will be seen by the OSS community. We should think about clarifying the benefits it brings and possibly bundling it together with the CompactionStrategyManager avoidance refactor.

I agree that these changes should either be bundled together with CSM refactor or otherwise seriously justified or... reverted during CEP preparation. However, I still think that porting them as-is as part of UCS and then dealing with them is easier then re-creating UCS based on the OSS code.

blambov commented on src/java/org/apache/cassandra/db/compaction/AbstractCompactionStrategy.java:
```diff
@@ -491,6 +499,37 @@ public synchronized void addSSTables(Iterable<SSTableReader> added)
+    /**
+     * Remove any tracked sstable that is no longer in the live set. Note that because we get notifications after the
+     * tracker is modified, anything we know of must be already in the live set -- if it is not, it has been removed
+     * from there, and we either we haven't received the removal notification yet, or we did and we messed it up (i.e.
```
Nit: There's one extra "we" here.
